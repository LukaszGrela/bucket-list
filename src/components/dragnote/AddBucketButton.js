import React from 'react';
import PropTypes from 'prop-types';

import './styles/AddBucketButton.scss'

export class AddBucketButton extends React.Component {

    clickHandler = () => {
        this.props && this.props.handleClick();
    }

    render = () => {
        return (
            <div className='add-bucket-button'>
                <button
                    onClick={this.clickHandler}>+</button>
            </div>
        )
    }
}

AddBucketButton.propTypes = {
    handleClick: PropTypes.func.isRequired
}

export default (AddBucketButton);
