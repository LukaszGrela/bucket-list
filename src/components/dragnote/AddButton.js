import React from 'react';
import PropTypes from 'prop-types';

import './styles/AddButton.scss'

export class AddButton extends React.Component {

    clickHandler = () => {
        this.props && this.props.handleClick();
    }

    render = () => {
        const { className } = this.props;
        return (
            <div className={'add-button' + (className ? ' ' + className : '')}>
                <button
                    onClick={this.clickHandler}>+</button>
            </div>
        )
    }
}

AddButton.propTypes = {
    handleClick: PropTypes.func.isRequired
}

export default (AddButton);
