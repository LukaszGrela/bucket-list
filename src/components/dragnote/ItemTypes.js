export const ItemTypes = {
    NOTE: 'dnd_di_note',
};