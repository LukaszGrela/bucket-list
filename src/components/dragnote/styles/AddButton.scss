.add-button {
    display: inline-block;
    overflow: hidden;
    button {
        width: 100%;
        height: 100%;
        cursor: pointer;
        padding: 1rem;
    }
}