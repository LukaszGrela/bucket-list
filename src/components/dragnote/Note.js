import React from 'react';
import PropTypes from 'prop-types';
import { DragSource } from 'react-dnd';
import { ItemTypes } from './ItemTypes';


import './styles/Note.scss';

class Note extends React.Component {
    render = () => {
        const { text, connectDragSource } = this.props;
        return connectDragSource(
            <div className='note'>{text}</div>
        )
    }
}
Note.propTypes = {
    id: PropTypes.string.isRequired,
    connectDragSource: PropTypes.func.isRequired,
    connectDragPreview: PropTypes.func.isRequired,
    isDragging: PropTypes.bool.isRequired,
};
const dragSource = {
    canDrag: (props) => {
        return true;
    },
    beginDrag: (props) => {
        console.log(props)
        return { id: props.id, bucketId: props.bucketId };
    }
}
const collect = (connect, monitor) => ({
    connectDragSource: connect.dragSource(),
    connectDragPreview: connect.dragPreview(),
    isDragging: monitor.isDragging(),
})
export default DragSource(ItemTypes.NOTE, dragSource, collect)(Note);