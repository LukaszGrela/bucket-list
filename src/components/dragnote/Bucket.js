import React from 'react';
import PropTypes from 'prop-types';
import Note from './Note';
import { DropTarget } from 'react-dnd';
import { ItemTypes } from './ItemTypes';
import './styles/Bucket.scss';
import AddWithText from '../AddWithText';



class Bucket extends React.Component {

    addNote = (note) => {
        const { addNoteHandler, id } = this.props;
        addNoteHandler && addNoteHandler(id, note);
    }
    render = () => {
        const { id, connectDropTarget,
            name, className,
            addNoteHandler, notes = { list: [] } } = this.props;
        const list = notes.list.filter(note => note.bucketId === id);
        return connectDropTarget(

            <div className={'bucket' + (className ? ' ' + className : '')}>
                <div className='bucket-content'>
                    <div className='bucket-name clearfix'>{name}</div>
                    <div className='notes-container clearfix'>
                        {
                            notes.loading ? "Notes loading..." :
                                list.map((note, index) => <Note key={note.id} {...note} />)

                        }
                    </div>
                    <div className='add-note clearfix'>
                        <AddWithText
                            placeholder={'Type a note'}
                            handleClick={this.addNote}
                        />
                    </div>
                </div>
            </div>
        )
    };
}
Bucket.propTypes = {
    addNoteHandler: PropTypes.func.isRequired,
    onDrop: PropTypes.func.isRequired,
    isOver: PropTypes.bool.isRequired,
    canDrop: PropTypes.bool.isRequired,
    connectDropTarget: PropTypes.func.isRequired,
};

const dropZoneTarget = {
    canDrop: (props, monitor) => {
        return true;
    },
    drop: (props, monitor) => {
        props.onDrop(props.id, monitor.getItem())
    }
};
const collect = (connect, monitor) => ({
    connectDropTarget: connect.dropTarget(),
    isOver: monitor.isOver(),
    canDrop: monitor.canDrop()
});

export default DropTarget(ItemTypes.NOTE,
    dropZoneTarget,
    collect)(Bucket);