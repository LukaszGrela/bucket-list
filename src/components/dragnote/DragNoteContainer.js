import React from 'react';
import { DragDropContext } from 'react-dnd';
import HTML5Backend from 'react-dnd-html5-backend';

import Bucket from './Bucket';
import { connect } from 'react-redux';
import { addBucket, swapBuckets } from '../../actions/actionBuckets';
import { addNote } from '../../actions/actionNotes';


import './styles/DragNoteContainer.scss';
import AddWithText from '../AddWithText';

class DragNoteContainer extends React.Component {
    addNote = (bucketId, note) => {
        /*
        TODO: show form modal
         */
        const { addNote } = this.props;
        addNote(bucketId, { text: note });
    }
    addBucket = (title) => {
        /*
        TODO: show input field for buckets
         */
        const { addBucket } = this.props;
        //create dummy bucket name
        addBucket(title);
    }
    onDrop = (bucketId, { id, bucketId: oldBucketId }) => {
        console.log(`Move note ${id} from bucket ${oldBucketId} to ${bucketId}`);
        const { swapBuckets } = this.props;
        swapBuckets(id, oldBucketId, bucketId);
    }
    render = () => {
        const { notes, buckets } = this.props;
        return (
            <div className='drag-note-container'>
                {/*<div>This is the DnD component, it will contain buckets and notes</div>
                <div>Add Bucket Button -> Input Bucket Form
                -> Cancel|OK -> dispatch addBucket
                -> firebase Rejected->Show Error Message|Fulfilled
                -> update store with new Bucket
                -> update view</div>*/}
                <div className='drag-note-container-wrapper'>
                    <div className='main-content'>
                        <div className='bucket-container'>
                            <div className='bucket-list'>
                                {
                                    buckets.list.map((bucket, index) => <Bucket
                                        className='item'
                                        key={bucket.id}
                                        {...bucket}
                                        notes={notes}
                                        onDrop={this.onDrop}
                                        addNoteHandler={this.addNote}
                                    />)
                                }
                                <div className='new-bucket'>
                                    <AddWithText
                                        placeholder={'Add new bucket'}
                                        handleClick={this.addBucket}
                                    />
                                </div>
                            </div>
                        </div>
                    </div>
                </div>
            </div>
        );
    }
}

const mapStateToProps = (state, props) => ({
    buckets: state.buckets,
    notes: state.notes
});
const mapDispatchToProps = (dispatch, props) => ({
    addBucket: (name) => dispatch(addBucket(name)),
    addNote: (bucketId, note) => dispatch(addNote(bucketId, note)),
    swapBuckets: (noteId, fromBucketId, toBucketId) => dispatch(swapBuckets(noteId, fromBucketId, toBucketId))
});
export default
    DragDropContext(HTML5Backend)(connect(
        mapStateToProps,
        mapDispatchToProps
    )(DragNoteContainer));