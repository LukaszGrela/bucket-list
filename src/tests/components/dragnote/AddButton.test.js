import React from 'react';
import { shallow } from 'enzyme';
import { AddButton } from '../../../components/dragnote/AddButton';


let wrapper, handleClick;
beforeEach(() => {
    handleClick = jest.fn();
    wrapper = shallow(<AddButton 
        handleClick={handleClick}/>);
});

test('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
});

test('should execute callback action', () => {
    const button = wrapper.find('.add-button button');

    expect(button).toHaveLength(1);

    button.simulate('click');

    expect(handleClick).toHaveBeenCalled();
})