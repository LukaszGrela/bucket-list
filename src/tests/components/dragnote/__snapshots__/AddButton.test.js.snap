// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div
  className="add-button"
>
  <button
    onClick={[Function]}
  >
    +
  </button>
</div>
`;
